docs/design/sql/08-expressions.md

```diff
@@ -0,0 +1,192 @@
+  `SqlOperator.getReturnTypeInference` and
+  `SqlOperator.inferReturnType`). The first one is used only if some of
+  the operand types are unknown. Also, operand type validation strategy
+  can be customized using an associated `SqlOperandTypeChecker` (see
+  `SqlOperator.getOperandTypeChecker`).
+
+- Additionally, some expressions require their operands to be coerced to
+  a certain type. For instance, `=` operator requires both of its sides
+  to be of the same type. In Calcite, such coercion services are
+  provided by `TypeCoercion` implementations. Specifically,
+  `TypeCoercionImpl` extended by `HazelcastTypeCoercion` is used to
+  customize the coercion logic.
+
+The following general rules apply when assigning types:
+
+- Literals: `TRUE` and `FALSE` literals receive `BOOLEAN` type; numeric
+  literals containing no decimal point (`1`, `42`, etc.) receive the
+  smallest integer type possible (`TINYINT`, `SMALLINT`, `INTEGER` or
+  `BIGINT`); numeric literals containing decimal point (`1.1`, `4.2`,
+  etc.) receive `DECIMAL` type; scientific notation numeric literals
+  (`1e1`, `4.2e2`, etc.) receive `DOUBLE` type; string literals
+  (`'foo'`) receive `VARCHAR` type.
+
+- Parameters: parameter types are inferred from the context: `1.0 + ?`,
+  the parameter would receive `DOUBLE` type.
+
+As an end result, for semantically valid ASTs, the validation process
+produces a potentially transformed AST where every node has a known
+type. For semantically invalid ASTs, an error is reported.
+
+### Conversion Stage
+
+The goal of the conversion stage it to convert the syntactically and
+semantically valid AST (`SqlNode`) received from the previous stages to
+a representation suitable for the relational optimization. In Calcite,
+relational nodes are represented by subclasses of `RelNode` while
+expression nodes referenced from relational ones are represented by
+subclasses of `RexNode`.
+
+During conversion Calcite also applies various simplifications and
+optimizations to expressions (see `SqlToRelConverter` and
+`HazelcastSqlToRelConverter`). After the conversion all references to
+the original textual SQL query representation are irreversibly lost. In
+other words, it's impossible to recover a `SqlNode` from a certain
+`RexNode` or `RelNode`.
+
+The resulting expressions may also undergo various transformations as
+dictated by relational optimizations applied in other parts of the SQL
+engine.
+
+### Translation Stage
+
+The goal of the translation phase is to translate the `RexNode`
+representation received from the previous stage to a representation
+suitable for the runtime evaluation of expressions.
+
+The translation is performed by `RexToExpressionVisitor` with the help
+of `RexToExpression`. Every instance of `RexNode` is translated into a
+corresponding `Expression` instance:
+
+The final result of the translation stage is an `Expression` tree ready
+for runtime evaluation. No further changes expected to the tree after
```

Comment:
No further changes _**are**_

Reply:
fixed in d1d2929413835f7a462bc88686630dd57203b27c